lib/AtomicCache.php

```diff
@@ -0,0 +1,174 @@
+    {
+        return call(function () use ($key, $create, $ttl): \Generator {
+            $value = yield $this->cache->get($key);
+
+            if ($value !== null) {
+                return $value;
+            }
+
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            try {
+                // Attempt to get the value again, since it may have been set while obtaining the lock.
+                $value = yield $this->cache->get($key);
+
+                if ($value !== null) {
+                    return $value;
+                }
+
+                return yield from $this->create($create, $key, null, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    /**
+     * The key is locked, the current value in the cache is accessed, then the $create callback is invoked with the key
+     * as the first parameter and the current value as the second parameter. The value returned from the callback is
+     * stored in cache and the promise returned from this method is resolved with the value.
+     *
+     * @param string   $key
+     * @param callable(string $key, mixed $value) $create
+     * @param int|null $ttl
+     *
+     * @return Promise<string>
+     *
+     * @throws CacheException If the $create callback throws an exception while generating the value.
+     */
+    public function swap(string $key, callable $create, ?int $ttl = null): Promise
+    {
+        return call(function () use ($key, $create, $ttl): \Generator {
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            $value = yield $this->cache->get($key);
+
+            try {
+                return yield from $this->create($create, $key, $value, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    private function lock(string $key): \Generator
+    {
+        try {
+            return yield $this->mutex->acquire($key);
+        } catch (\Throwable $exception) {
+            throw new CacheException(
+                \sprintf('Exception thrown when obtaining the lock for key "%s"', $key),
+                0,
+                $exception
+            );
+        }
+    }
+
+    private function create(callable $create, string $key, ?string $value, ?int $ttl = null): \Generator
+    {
+        try {
+            $value = yield call($create, $key, $value);
+        } catch (\Throwable $exception) {
+            throw new CacheException(
+                \sprintf('Exception thrown while creating the value for key "%s"', $key),
+                0,
+                $exception
+            );
+        }
+
+        if ($value === null) {
+            throw new CacheException(\sprintf('The value to be cached for key "%s" must not be NULL', $key));
+        }
+
+        $value = (string) $value;
```

Comment:
Personally I'd rather force the callback to return string.

```suggestion
        if (! is_string($value)) {
            throw new CacheException(\sprintf('The value to be cached for key "%s" must be a string', $key));
        }
```

Reply:
Agreed and updated.